conda/core/index.py

```diff
@@ -51,7 +58,457 @@ def check_allowlist(channel_urls: list[str]) -> None:
 LAST_CHANNEL_URLS = []
 
 
+class Index(UserDict):
+    """The ``Index`` provides information about available packages from all relevant sources.
+
+    There are four types of sources for package information, namely
+
+    Channels
+        represent packages available from standard sources identified with a url, mostly online,
+        but can also be on a local filesystem using the `file://` scheme.
+        Programatically, channels are represented by :class:`conda.models.Channel`, their data
+        is fetched using :class:`conda.index.subdir_data.SubdirData`.
+
+        For more information see :ref:`condepts-channels`.
+
+    Prefix
+        represents packages that are already installed. Every :class:`Index` can be associated
+        with exactly one Prefix, which is the location of one of the conda :ref:`concepts-environments`.
+        The package information about the installed packages is represented by :class:`conda.prefix.prefix_data.PrefixData`.
+
+    Package Cache
+        represents packages that are locally unpacked, but may not be installed in the environment
+        associated with this index. These are usually packages that have been installed in any environment
+        of the local conda installation, but may have been removed from all environments by now.
+
+    Virtual Packages
+        represent properties of the system, not actual conda packages in the normal sense. These are,
+        for example, system packages that inform the solver about the operating system in use, or
+        track features that can be used to steer package priority.
+    """
+
+    def __init__(
+        self,
+        channels=(),
+        prepend=True,
+        platform=None,
+        subdirs=None,
+        use_local=False,
+        use_cache=None,
+        prefix=None,
+        repodata_fn=context.repodata_fns[-1],
+        add_system=False,
+    ) -> None:
+        if use_local:
+            channels = ["local"] + list(channels)
+        if prepend:
+            channels += context.channels
```

Comment:
Is this prepending or appending? 🤔

Reply:
Oh, good point that's.. not ideal variable naming, the code currently appends.

Reply:
So this had me confused as well. A little explanation: Currently, `get_index` produces a simple dictionary, which it does by going through the list of channels and adding *or overwriting* packages. That means that channels later in the list effectively have higher priority for equal packages. In the new code this is done by traversing the list of channels in `__getitem__` in reversed order (l.247).

Effectively this code is inlining `calculate_channel_urls` and `all_channel_urls`.